packages/flutter/lib/src/services/text_input.dart

```diff
@@ -704,6 +722,387 @@ class RawFloatingCursorPoint {
+
+      if (!isDeletionGreaterThanOne) {
+        actualStart = end - 1;
+      }
+
+      return TextEditingDeltaDeletion(
+        oldText: oldText,
+        deltaText: tb,
+        deltaRange: TextRange(
+          start: actualStart,
+          end: end,
+        ),
+        selection: newSelection,
+        composing: newComposing,
+      );
+    } else if ((start == end || isInsertingInsideComposingRegion) &&
+        !isOriginalComposingRegionTextChanged) {  // Insertion.
+      return TextEditingDeltaInsertion(
+        oldText: oldText,
+        deltaText: tb.substring(end - start, (end - start) + (tb.length - (end - start))),
+        deltaRange: TextRange(
+          start: end,
+          end: end,
+        ),
+        selection: newSelection,
+        composing: newComposing,
+      );
+    } else if (isReplaced) {  // Replacement.
+      return TextEditingDeltaReplacement(
+        oldText: oldText,
+        deltaText: tb,
+        deltaRange: TextRange(
+          start: start,
+          end: end,
+        ),
+        selection: newSelection,
+        composing: newComposing,
+      );
+    }
+
+    assert(false);
+    return TextEditingDeltaNonTextUpdate(
+      oldText: oldText,
+      selection: newSelection,
+      composing: newComposing,
+    );
+  }
+
+  /// The old text state before the delta has occurred.
+  final String oldText;
+
+  /// The value represents the text that is being inserted/deleted by this delta.
+  ///
+  /// This value will slightly vary based on the [TextEditingDeltaType].
+  ///
+  /// For a [TextEditingDeltaType.insertion] this will be the character/s being
+  /// inserted.
+  ///
+  /// For a [TextEditingDeltaType.deletion] this will be an empty string.
+  ///
+  /// For a [TextEditingDeltaType.replacement] this will be the text that is
+  /// replacing the [TextEditingDelta.deltaRange].
+  ///
+  /// For a [TextEditingDeltaType.nonTextUpdate] this will be an empty string.
+  final String deltaText;
+
+  /// This value can either represent a range of text that the delta is changing
+  /// or if is a collapsed range then it represents the point where this delta
+  /// began.
+  ///
+  /// This value will slightly vary based on the [TextEditingDeltaType].
+  ///
+  /// For a [TextEditingDeltaType.insertion] this will be a collapsed range
+  /// representing the cursor position where the insertion began.
+  ///
+  /// For a [TextEditingDeltaType.deletion] this will be the range of text
+  /// that was deleted.
+  ///
+  /// For a [TextEditingDeltaType.replacement] this will be the range of
+  /// characters that are being replaced.
+  ///
+  /// For a [TextEditingDeltaType.nonTextUpdate] this will be a collapsed range
+  /// of (-1,-1).
+  final TextRange deltaRange;
+
+  /// {@template flutter.services.TextEditingDelta.deltaType}
+  /// The type of delta that has occured.
+  ///
+  /// See [TextEditingDeltaType] for more information.
+  /// {@endtemplate}
+  TextEditingDeltaType get deltaType;
```

Comment:
What should users do with deltaType if they implement their own TextEditingDelta subclass?

Reply:
This was discussed offline but @LongCatIsLooong suggested converting any new types to one of the basic types we have introduced `replacement/deletion/insertion/nonTextUpdate`. If a developer does want to implement their own `TextEditingDelta` subclass I think that would be the appropriate course of action if they want to avoid any framework changes. 

Though how a developer would plug this new delta type into the current pipeline has not been addressed. Currently that is all handled by the `TextInput._handleTextInputInvocation` and all the developer will ever receive is the final list of `TextEditingDeltas`. In theory they could process this list of `TextEditingDeltas` even further to include their own custom `TextEditingDelta` types, but there is no way to include their new delta type in the processing before this point.